torch/csrc/api/include/torch/nn/cloneable.h

```diff
@@ -48,9 +48,10 @@ class Cloneable : public virtual Module {
         "Are you sure you called register_parameter() inside reset() "
         "and not the constructor?");
     for (const auto& parameter : parameters_) {
-      auto data = autograd::Variable(*parameter).clone();
-      copy->parameters_[parameter.key()].set_data(
-          device ? data.to(*device) : data);
+      auto data = device ?
+          (*parameter).to(*device) :
```

Comment:
Thanks for the fix. However, I'm not so sure about changing `.clone`. The name suggests that it should perform a deepcopy. Does c++ api modules have a `.to`? IMO, it should have one and with optional argument `bool copy=false`, just like the one in python. If we have a choice, I would prefer adding a `.to` and deprecating `.clone` for it.

Reply:
@SsnL  C++ `module` does have a `.to` API, but it does not support `bool copy` argument, and will modify the module's own parameters instead of creating a new module. It seems Python's [`module.to`](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L311) does not support `bool copy` either. Am I looking at the wrong place?

IIUC, `Tensor.to`will copy instead of move, and hence `Module.clone()` still performs a deepcopy after this PR, no?

Reply:
Ahh yes you are totally right! `.to` is inplace. 

So.. hmm how about add a bool flag to `.clone`? Maybe named `deep` or `force_copy` etc? Idk...

Reply:
Sure, I can make that change. Let me make sure I understand the requirement. Are you suggesting that for `Module.clone()`, we are going to support two modes:

1. `deep=True` will be same as current `Module.clone()`'s behavior, which always make a deep copy, i.e., `(*parameter).clone()` if on the same device and `(*parameter).to(device)` otherwise.
2. `deep=False`, will only make a copy when cloning to a different device.

> I'm not so sure about changing .clone. The name suggests that it should perform a deepcopy. 

I think I misunderstood your comments yesterday. There was indeed a problem in that version when `device` is provided but is the same as the current device, where `Tensor.to(device)` becomes inplace. Fixed in the new commit. It seems some `ModuleTest` in `module.cpp` are disabled, let me check.